### PR TITLE
Update from SDKv2 to terraform-framework

### LINKED ISSUE #328: Storage pool configuration inconsistencies
This issue is intended more as a discussion (or question) then an actual issue.

---

When creating a storage pool, specific configurations must be provided depending on the storage pool driver in use. If not, reapplying the identical Terraform configuration leads to an error.

## Steps to reproduce

Terraform configuration:
```hcl
resource "lxd_storage_pool" "pool" {
  name    = "tf-pool"
  driver  = "dir"
}
```

Commands:
```sh
$ terraform apply --auto-approve
$ terraform apply # Reapply the same configuration

Terraform will perform the following actions:

  # lxd_storage_pool.pool will be updated in-place
  ~ resource "lxd_storage_pool" "pool" {
      ~ config = {
          - "source" = "/var/snap/lxd/common/lxd/storage-pools/tf-pool" -> null
        }
        id     = "tf-pool"
        name   = "tf-pool"
        # (1 unchanged attribute hidden)
    }
```

The error `Error: Pool source cannot be changed when not in pending state` arises because the provider attempts to reapply an empty configuration.

Although the `source` can be auto-determined based on the LXD installation location and the storage pool driver in use (except for Ceph), different drivers rely on unique configuration settings. For instance, `zfs` sets `zfs.size` and `zfs.pool_name`. This can lead to inconsistencies even if `source` is auto-determined (which would still be nice to have).

The core problem is that applying the identical Terraform configuration multiple times produces different outcomes.

## Potential solutions

Just brainstorming...

### Update keys rather than replacing the configuration

One approach might be to update or add keys in the existing configuration instead of replacing the entire configuration. However, this has a significant downside: keys can't be removed unless they're explicitly set to an empty string in the `config` block. This means any user-added keys can't be undone, making the Terraform state unreliable.

### Set of required keys for each storage pool driver

We could specify required configuration keys for each type of storage pool. This ensures that the `config` block can be safely applied, and the same Terraform configuration would produce consistent results when reapplied.

---

Any thoughts or suggestions?

### LINKED ISSUE #380: deprecate remaining container attributes
`lxd_publish_image` and `lxd_snapshot` resources still use the legacy names `container` and `container_name`

These attributes need to be deprecated in favor of `instance` attributes. 

This work may want to wait on #381

### LINKED ISSUE #381: Upgrade to Terraform Plugin Framework
https://developer.hashicorp.com/terraform/plugin/framework

### PR DESCRIPTION
**Pull request by MusicDin**
Main contributions of this PR:
- Changes from SDKv2 to terraform-framework 
  - All new resources are in `/internal`, separated into multiple packages, which means if changing the structure is just about moving the files around.  
  - Old SDKv2 files are left untouched in `/lxd` (except `main.go` file)
- Fixes configuration inconsistencies
  - Resources that use config files have `ComputedKeys` function that returns a list of computed keys for that resource (e.g. storage pool).
  - Computed keys are removed from the config on Read (if the user has not defined them). 
  - On Update, user configuration is merged with the resource's config. During merge, user keys are retained and computed keys are appended (if found in fetched resource config). All other keys are removed, in order to ensure the desired infrastructure state.

---

TODO:

- [x] **Computed attributes**
  - ~Currently, there is too many computed attributes. This is my bad, and will fix this soon. It does not change the functionality, just makes the resource attributes (in HCL) accessible to other resources.~
  - I've just learned that all fields that are set within a resource can be accessed from other resources, so this is not an issue.

- [x] **Docs**:
  - [x] Document computed keys for each resource that uses them.
  - [x] Documentation consistency. It would be nice if we have some rule how attributes are ordered in the documentation, to make it easier to read and spot mistakes. 

- [x] **GH Workflow**. 
  - [x] Call the new tests in test workflow.
  - [x] Prechecks are a bit tricky now, because we have no access to the provider instance within the tests. See `Tests` in backlog. As a temporary solution, we use default lxd configuration which defines `local` remote (assuming all tests are run against local lxd).
  
TODO (backlog):

**Tests**: Framework uses new approach with `ProviderFactories`. Due to this, I was unable to access the initialized provider within the tests ([here](https://discuss.hashicorp.com/t/terraform-framework-access-provider-within-tests/60101) is a question on terraform community forum).
  - LXD API is not called after each test to verify resource existence. For now, we can rely on computed keys, such as `status` = `Running` for instances.

Questions:
- Can we agree upon one approach for limits?
  - `Profile` resource tracks limits in `config` attribute (`config = { "limits.cpu" = 2 }`)
  - `Instance` resource tracks limits in `limits` attribute (`limits = { "cpu" = 2 }`)

Fixes #328
Fixes #380
Fixes #381
Closes #376

### CODE REVIEW

**Comment by MusicDin** [comment]
New tests seem to pass except file upload on instance where file `mode` is flip/flopping between the configured value and default value. Seems that this is an ongoing issue https://github.com/hashicorp/terraform-plugin-framework/issues/867

**Comment by MusicDin** [comment]
Even though tests are currently passing, there seems to be an issue with images which I cannot reproduce locally. It happens only on GH actions, and even there just occasionally.

Example:
```
Error: Failed to create instance "subtle-python"

Failed instance creation: Failed creating instance from image: Image is
missing a rootfs:
/var/snap/lxd/common/lxd/images/d1c496f8540f889dacbfa1a019fb31c31e93aaca649f3bcb670e49ec1476c0a0
```

Some links to action logs which contain this error:
- https://github.com/terraform-lxd/terraform-provider-lxd/actions/runs/7016285354/job/19087178846#step:8:126
- https://github.com/terraform-lxd/terraform-provider-lxd/actions/runs/6984741961/job/19008031826#step:8:220
- https://github.com/terraform-lxd/terraform-provider-lxd/actions/runs/6984741961/job/19008031715#step:8:332

**Comment by MusicDin** [comment]
Image issue should be fixed now. Seems that tests from `lxd_cached_image` removed cached image while other tests were running, causing random failures and error messages. So the solution is to use different images for testing cached_image resource, until this is fixed on LXD side.

There will be no more changes to this PR (unless there are any comments), so it is ready to be reviewed/merged.

File: docs/resources/volume.md [hunk 1]
```diff
@@ -6,48 +6,94 @@ Manages an LXD volume.
...
+
+Import ID syntax: `[<remote>:][<project>]/<pool>/<name>[,content_type=<ct>]`
+
+* `<remote>` - *Optional* - Remote name.
+* `<project>` - *Optional* - Project name.
+* `<pool>` - **Required** - Storage pool name.
+* `<name>` - **Required** - Volume name.
+* `content_type=<ct>` - *Optional* - Content type of the volume.
+
+-> Currently, only volumes of type `custom` can be imported.
```
[the remark this answers is not in the document]
Comment (MusicDin) [reply docs/resources/volume.md:60]: Volume can be of `type` custom, but its `content_type` can be filesystem/block. I didn't know how to put this together without writing a long paragraph haha

File: docs/resources/instance.md [hunk 2]
```diff
@@ -185,20 +181,43 @@ resource "lxd_instance" "instance1" {
...
-Existing instances can be imported with the following syntax:
+Import ID syntax: `[<remote>:][<project>/]<name>[,image=<image>]`
+
+* `<remote>` - *Optional* - Remote name.
+* `<project>` - *Optional* - Project name.
+* `<name>` - **Required** - Instance name.
+* `image=<image>` - *Optional* - The image used by the instance.
+
+~> **Warning:** Importing the instance without specifying `image` will lead to its replacement
+   upon the next apply, rather than an in-place update.
```
[the remark this answers is not in the document]
Comment (MusicDin) [reply docs/resources/instance.md:191]: Idk. I need to check, but I think I've fixed that, so this image should not be required at all.

However, we (currently) do not support empty instances.

[the remark this answers is not in the document]
Comment (MusicDin) [reply docs/resources/instance.md:191]: So the issue here is that we cannot extract the image name (e.g. `images:alpine/3.18/amd64`) when reading the instance. We can get the fingerprint, but it does not help much in this situation.

**Review by adamcstephens — COMMENTED** [review]
This looks pretty good to me, but given the size of the diff I'm mostly relying on the tests and manually running against my environment.

The one issue I've run into is that instances want to be replaced due to some changes in cluster targeting. Any ideas on how to prevent this replacement?

```
-/+ resource "lxd_instance" "guide2" {
      ~ ipv4_address     = "10.0.0.10" -> (known after apply)
      ~ mac_address      = "00:16:3e:15:15:23" -> (known after apply)
        name             = "guide2"
      ~ status           = "Running" -> (known after apply)
      - target           = "none" -> null # forces replacement
```

**Comment by MusicDin** [comment]
@adamcstephens Thanks for the review. 

> I'm good with target_path and source_path. I was going to suggest target/source, but we already use target for the cluster targeting.

I'll change file target/source path attributes.

> The one issue I've run into is that instances want to be replaced due to some changes in cluster targeting. Any ideas on how to prevent this replacement?

Need to check this one for sure, but I think this is because the target is being set to `null` if value is either `""` or `none`, and it triggers a replacement if value changes.
This has to be fixed (on all resources that use target) regardless of the reason.

**Comment by adamcstephens** [comment]
On this standalone server I've never specified a cluster target, so it's receiving the previous default. Is there a way we can migrate this from `"none"` to `null` for the user so they don't have to fix this manually?

**Comment by MusicDin** [comment]
I'm just looking into this. I think we can modify the plan and conditionally remove `RequiresReplace` plan modifier if the target is `none`.

I'm just curious what happens if someone's target is actually named `none`.. Does it make sense to do this only if server is not in cluster mode?

**Comment by MusicDin** [comment]
Now, we always set target to `""`, and only replace the resource, if user actually configured the value.

@adamcstephens Can you please verify that `none` gets rewritten to the `""` without instance being replaced?

**Comment by adamcstephens** [comment]
@MusicDin I can confirm that the changes fix the issue for me. Instances are no longer targeted for replacement.

**Review by adamcstephens — APPROVED** [review]
Thanks for this. Looks like a non-trivial amount of effort, but it'll be nice to finally be up to date with the recommended provider setup.

**Review by simondeziel — APPROVED** [review]
LGTM, thanks again!

**Pull request merged**